src/wallet/rpcdump.cpp

```diff
@@ -350,6 +350,98 @@ UniValue importaddress(const JSONRPCRequest& request)
     return NullUniValue;
 }
 
+static void RemoveAddress(CWallet*, const CTxDestination& dest);
+static void RemoveScript(CWallet* const pwallet, const CScript& script, bool isRedeemScript) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+{
+    isminetype ismine = ::IsMine(*pwallet, script);
+    if (!isRedeemScript && ismine == ISMINE_SPENDABLE) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "The wallet contains the private key for this address");
+    } else if (!isRedeemScript && ismine == ISMINE_NO) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "The wallet does not contain this address or script");
+    }
+    assert(ismine == ISMINE_WATCH_ONLY || isRedeemScript);
+
+    if (pwallet->HaveWatchOnly(script) && !pwallet->RemoveWatchOnly(script)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error removing address/script from wallet");
+    }
+    if (isRedeemScript){
+        const CScriptID id(script);
+        if (!pwallet->HaveCScript(id)) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Error removing p2sh redeemScript from wallet");
+        }
+        RemoveAddress(pwallet, id);
+    } else {
+        CTxDestination destination;
+        if (ExtractDestination(script, destination)) {
+            pwallet->DelAddressBook(destination);
+        }
+    }
+    pwallet->RemoveOldTransactions();
+}
+
+static void RemoveAddress(CWallet* const pwallet, const CTxDestination& dest) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+{
+    CScript script = GetScriptForDestination(dest);
+    RemoveScript(pwallet, script, false);
+    // remove from address book
+    if (IsValidDestination(dest)) {
+        pwallet->DelAddressBook(dest);
+    }
+}
+
+UniValue removeaddress(const JSONRPCRequest& request)
+{
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    CWallet* const pwallet = wallet.get();
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            RPCHelpMan{"removeaddress",
+                "\nRemoves an address or script (in hex) that was being watched as if it were in your wallet but was not being used to spend. Requires a new wallet backup.\n",
+                {
+                    {"address", RPCArg::Type::STR, /* opt */ false, /* default_val */ "", "The Bitcoin address (or hex-encoded script)"},
+                    {"p2sh", RPCArg::Type::BOOL, /* opt */ true, /* default_val */ "false", "Add the P2SH version of the script as well"},
```

Comment:
Did you mean "Remove" instead of "Add"?

Reply:
Suggest adding a third argument `purge_transactions` (default true) about removing existing transactions from the wallet.